plugins/experimental/fq_pacing/fq_pacing.c

```diff
@@ -0,0 +1,199 @@
+TSReturnCode
+TSRemapInit(TSRemapInterface *api_info, char *errbuf, int errbuf_size)
+{
+  if (!api_info) {
+    strncpy(errbuf, "[fq_pacing] - Invalid TSRemapInterface argument", (size_t)(errbuf_size - 1));
+    return TS_ERROR;
+  }
+
+  if (api_info->size < sizeof(TSRemapInterface)) {
+    strncpy(errbuf, "[TSRemapInit] - Incorrect size of TSRemapInterface structure", errbuf_size - 1);
+    return TS_ERROR;
+  }
+
+  if (api_info->tsremap_version < TSREMAP_VERSION) {
+    snprintf(errbuf, errbuf_size - 1, "[TSRemapInit] - Incorrect API version %ld.%ld", api_info->tsremap_version >> 16,
+             (api_info->tsremap_version & 0xffff));
+    return TS_ERROR;
+  }
+
+  TSDebug(PLUGIN_NAME, "plugin is succesfully initialized");
+  return TS_SUCCESS;
+}
+
+TSReturnCode
+TSRemapNewInstance(int argc, char *argv[], void **ih, char *errbuf, int errbuf_size)
+{
+  fq_pacing_cfg_t *cfg      = NULL;
+  unsigned long pacing_rate = 0;
+
+  TSDebug(PLUGIN_NAME, "Instantiating a new remap.config plugin rule");
+
+  if (argc > 1) {
+    int c;
+    static const struct option longopts[] = {{"rate", required_argument, NULL, 'r'}, {NULL, 0, NULL, 0}};
+
+    // The "-" in optstring is required to prevent permutation of argv, which
+    // makes the plugin loader crashy
+    while ((c = getopt_long(argc, (char *const *)argv, "-r:", longopts, NULL)) != -1) {
+      switch (c) {
+      case 'r':
+        errno       = 0;
+        pacing_rate = strtoul(optarg, NULL, 0);
+        if (errno != 0) {
+          snprintf(errbuf, errbuf_size - 1, "[TsRemapNewInstance] input pacing value is not a valid positive integer");
+          return TS_ERROR;
+        }
+
+        break;
+      }
+    }
+  }
+
+  if (pacing_rate > MAX_PACING_RATE) {
+    snprintf(errbuf, errbuf_size - 1, "[TsRemapNewInstance] input pacing value is too large (%lu), max(%lu)", pacing_rate,
+             MAX_PACING_RATE);
+    return TS_ERROR;
+  }
+
+  cfg = TSmalloc(sizeof(fq_pacing_cfg_t));
+  memset(cfg, 0, sizeof(*cfg));
+  cfg->pacing_rate = pacing_rate;
+  *ih              = (void *)cfg;
+  TSDebug(PLUGIN_NAME, "Setting pacing rate to %lu", pacing_rate);
+
+  return TS_SUCCESS;
+}
+
+void
+TSRemapDeleteInstance(void *instance)
+{
+  TSError("[fq_pacing] Cleaning up...");
+
+  if (instance != NULL) {
+    TSfree((fq_pacing_cfg_t *)instance);
+  }
+}
+
+static int
+reset_pacing_cont(TSCont contp, TSEvent event, void *edata)
+{
+  TSHttpTxn txnp             = (TSHttpTxn)edata;
+  fq_pacing_cont_t *txn_data = TSContDataGet(contp);
+
+#ifdef SO_MAX_PACING_RATE
+  if (txn_data->client_fd > 0) {
+    TSDebug(PLUGIN_NAME, "Disabling SO_MAX_PACING_RATE for client_fd=%d", txn_data->client_fd);
+    int res = 0;
+    res     = safe_setsockopt(txn_data->client_fd, SOL_SOCKET, SO_MAX_PACING_RATE, (char *)&pacing_off, sizeof(pacing_off));
+    // EBADF indicates possible client abort
+    if ((res < 0) && (errno != EBADF)) {
+      TSError("fq_pacing] Error disabling SO_MAX_PACING_RATE, errno=%d", errno);
```

Comment:
Tiniest of all possible nitpicks... missing a `[` at the start of this string.

Reply:
Also, if fq is not enabled, this error will be spammed once per connection in the logs, right? That might be a bit much.

Reply:
Will fix the missing bracket. 

I'll look at adding some code at plugin init time to read /proc/sys/net/core/default_qdisc and fail initialization if its not 'fq'